File: ActionUI/Views/Table.swift
// Sources/Views/Table.swift
/*
 Sample JSON for Table view (macOS only):
 {
   "type": "Table",
   "id": 1,              // Required: Non-zero positive integer for runtime programmatic interaction and diffing
   "properties": {
     "columns": ["Name", "Action", "Icon"], // Required: Array of strings for column headers
     "columnTypes": [                       // Optional: Per-column type config array. Defaults to all Text.
       { "viewType": "Text" },              // Each entry: { "viewType": "Text"|"Button"|"Image"|"AsyncImage"
       { "viewType": "Button",              // Columns without an entry default to Text.
         "actionContext": "rowIndex",       // "actionContext": "title"|"rowIndex"|"columnIndex"|"rowColumnIndex" (Button only)
         "actionID": "row.action" },        // "actionID": "..." (Button only — fires on button click) }
       { "viewType": "Image",
         "dataInterpretation": "systemName" } // "dataInterpretation": "path"|"systemName"|"assetName"|"resourceName"|"mixed" (Image only)
     ],
     "widths": [100, 80, 40],               // Optional: Array of integers for ideal column widths (resizable; last column fills remaining space)
     "actionID": "table.selection.changed", // Optional: Fires on selection change (all cell types)
     "doubleClickActionID": "table.double.click" // Optional: String for double-click action (context = row index)
   }
 }
   // Note: The Table view is macOS-only, showing a multi-column table with per-column cell types specified by the columnTypes array. If columnTypes is omitted or shorter than columns, missing entries default to Text. Selection is stored as [String] in state["value"], using row IDs for tracking. The table-level actionID fires on selection change. Button columns have their own actionID in their columnTypes entry, fired on click — this cleanly separates selection events from button click events. Baseline View properties (padding, hidden, foregroundColor, font, background, frame, opacity, cornerRadius, actionID, disabled) and additional View protocol modifiers are inherited and applied via ActionUIRegistry.shared.applyViewModifiers(to: baseView, properties: element.properties). The applyModifiers implementation is provided by the ActionUIViewConstruction protocol extension. SwiftUI types are explicitly prefixed (e.g., SwiftUI.Table, SwiftUI.TableColumn) to avoid namespace conflicts. Uses TableColumnForEach for dynamic columns on macOS 14.4+. Falls back to a placeholder message for earlier versions.
   // Performance: Child views are strongly typed to avoid AnyView overhead, identified by stable indices in ForEach, optimizing SwiftUI diffing for large tables (e.g., 1000 rows x 50 columns). Image creation uses SwiftUI.Image extension, aligned with Image.swift, to minimize overhead. Ensure state updates are targeted to minimize re-renders.

 Observable state:
   value ([String])                    Selected row as an array of column strings (first column = display value).
                                       Access via getElementValue / setElementValue.
   states["content"]   [[String]]      All table rows; each inner array holds one row's column values.
                                       Access via getElementRows / setElementRows / appendElementRows /
                                       clearElementRows / getElementColumnCount.
   states["selectedRowID"] String?     Stable row ID of the currently selected row; nil when nothing is
                                       selected. No dedicated public API — use getElementState / setElementState.
 */

import SwiftUI

struct TableRowData: Identifiable {
    let id: String
    let values: [String]
}

struct ColumnData: Identifiable {
    let id: Int
    let name: String
    let minWidth: CGFloat?
    let idealWidth: CGFloat?
    let maxWidth: CGFloat?
}

struct Table: ActionUIViewConstruction {
    static let valueType: Any.Type = [String].self // Value is the selected row as [String]
    
    static var validateProperties: ([String: Any], any ActionUILogger) -> [String: Any] = { properties, logger in
        var validatedProperties = properties
        
        // Parse columnTypes array — each entry: { viewType, dataInterpretation?, actionContext?, actionID? }
        var columnTypes = properties["columnTypes"] as? [[String: Any]] ?? []
        let columns = properties["columns"] as? [String] ?? []
        // Pad to match columns count, defaulting to Text
        while columnTypes.count < columns.count {
            columnTypes.append(["viewType": "Text"])
        }
        // Validate each entry
        for i in 0..<columnTypes.count {
            var ct = columnTypes[i]
            let vt = ct["viewType"] as? String ?? "Text"
            if !["Text", "Button", "Image", "AsyncImage"].contains(vt) {
                logger.log("Table columnTypes[\(i)].viewType must be 'Text', 'Button', 'Image', or 'AsyncImage'; defaulting to Text", .warning)
                ct["viewType"] = "Text"
            }
            if vt == "Image" {
                let di = ct["dataInterpretation"] as? String
                if !["path", "systemName", "assetName", "resourceName", "mixed"].contains(di) {
                    logger.log("Table columnTypes[\(i)].dataInterpretation must be 'path', 'systemName', 'assetName', 'resourceName', or 'mixed' for Image; defaulting to systemName", .warning)
                    ct["dataInterpretation"] = "systemName"
                }
            }
            if vt == "Button" {
                let ac = ct["actionContext"] as? String
                if !["title", "rowIndex", "columnIndex", "rowColumnIndex"].contains(ac) {
                    logger.log("Table columnTypes[\(i)].actionContext must be 'title', 'rowIndex', 'columnIndex', or 'rowColumnIndex' for Button; defaulting to title", .warning)
                    ct["actionContext"] = "title"
                }
            }
            columnTypes[i] = ct
        }
        validatedProperties["columnTypes"] = columnTypes
        
        if validatedProperties["columns"] == nil {
            validatedProperties["columns"] = []
        } else if !(validatedProperties["columns"] is [String]) {
            logger.log("Table columns must be an array of strings; defaulting to []", .warning)
            validatedProperties["columns"] = []
        }
        if let widths = properties["widths"] as? [Int] {
            validatedProperties["widths"] = widths
        } else if properties["widths"] != nil {
            logger.log("Table widths must be an array of integers; ignoring", .warning)
            validatedProperties["widths"] = nil
        }

        if let doubleClickActionID = properties["doubleClickActionID"] as? String {
            validatedProperties["doubleClickActionID"] = doubleClickActionID
        } else if properties["doubleClickActionID"] != nil {
            logger.log("Table doubleClickActionID must be a string; ignoring", .warning)
            validatedProperties["doubleClickActionID"] = nil
        }
        
        return validatedProperties
    }
    
    static var initialStates: (ViewModel) -> [String: Any] = { model in
        var states: [String: Any] = model.states
        if states.isEmpty {
            states["content"] = [] as [[String]]
            states["selectedRowID"] = nil
        }
        return states
    }
    
    static var buildView: (any ActionUIElementBase, ViewModel, String, [String: Any], any ActionUILogger) -> any SwiftUI.View = { element, model, windowUUID, properties, logger in
        #if canImport(AppKit)
        let columnTypes = properties["columnTypes"] as? [[String: Any]] ?? []
        let columns = (properties["columns"] as? [String]) ?? []
        let rows = (model.states["content"] as? [[String]]) ?? []
        let idealWidths = (properties["widths"] as? [Int])?.map { CGFloat($0) }
        let lastVisibleIndex = columns.count - 1

        let columnData = columns.enumerated().map { (index, name) in
            let ideal: CGFloat? = idealWidths.flatMap { $0.indices.contains(index) ? $0[index] : nil } ?? 100
            let isLast = (index == lastVisibleIndex)
            return ColumnData(
                id: index,
                name: name,
                minWidth: 40,
                idealWidth: ideal,
                maxWidth: isLast ? .infinity : nil
            )
        }
        let rowData = rows.enumerated().map { (index, row) in
            TableRowData(id: "row-\(index)", values: row)
        }
        
        let selectionBinding = Binding<Set<String>>(
            get: {
                guard let selectedRow = model.value as? [String],
                      !selectedRow.isEmpty else {
                    return Set<String>()
                }
                // Find which row-id corresponds to this value array
                if let matchingRow = rowData.first(where: { $0.values == selectedRow }) {
                    return Set([matchingRow.id])
                }
                return Set<String>()
            },
            set: { newSet in
                let newRowID: String? = newSet.first   // enforce single for now
        
                var selectedRowValues: [String] = []
        
                if let rowID = newRowID,
                   let selectedRow = rowData.first(where: { $0.id == rowID }) {
                    selectedRowValues = selectedRow.values
                }
        
                guard (model.value as? [String]) != selectedRowValues else { return }
        
                DispatchQueue.main.async {
                    model.value = selectedRowValues
        
                    if let actionID = properties["actionID"] as? String {
                        ActionUIModel.shared.actionHandler(
                            actionID,
                            windowUUID: windowUUID,
                            viewID: element.id,
                            viewPartID: 0
                        )
                    }
                }
            }
        )
        
        return SwiftUI.Table(rowData, selection: selectionBinding) {
            SwiftUI.TableColumnForEach(columnData) { column in
                SwiftUI.TableColumn(column.name) { row in
                    let value = column.id < row.values.count ? row.values[column.id] : ""
                    let colType = column.id < columnTypes.count ? columnTypes[column.id] : ["viewType": "Text"]
                    let viewType = colType["viewType"] as? String ?? "Text"
                    let dataInterpretation = colType["dataInterpretation"] as? String ?? "systemName"
                    let actionContext = colType["actionContext"] as? String ?? "title"
                    SwiftUI.Group {
                        switch viewType {
                        case "Text":
                            SwiftUI.Text(value)
                        case "Button":
                            SwiftUI.Button(value) {
                                if let buttonActionID = colType["actionID"] as? String {
                                    let context: Any = {
                                        switch actionContext {
                                        case "rowIndex": return rowData.firstIndex(where: { $0.id == row.id }) ?? -1
                                        case "columnIndex": return column.id
                                        case "rowColumnIndex": return Point(row: rowData.firstIndex(where: { $0.id == row.id }) ?? -1, column: column.id)
                                        default: return value
                                        }
                                    }()
                                    ActionUIModel.shared.actionHandler(buttonActionID, windowUUID: windowUUID, viewID: element.id, viewPartID: column.id, context: context)
                                }
                            }
                        case "Image":
                            SwiftUI.Image(from: value, interpretation: dataInterpretation)
                        case "AsyncImage":
                            SwiftUI.AsyncImage(url: URL(string: value)) { image in
                                image.resizable().scaledToFit()
                            } placeholder: {
                                SwiftUI.ProgressView()
                            }
                        default:
                            SwiftUI.Text(value)
                        }
                    }
                }
                .width(min: column.minWidth, ideal: column.idealWidth, max: column.maxWidth)
            }
        }
        .onTapGesture(count: 2) {
            if let doubleClickActionID = properties["doubleClickActionID"] as? String,
               let selectedRow = model.value as? [String],
               !selectedRow.isEmpty,
               let index = rowData.firstIndex(where: { $0.values == selectedRow }) {
                ActionUIModel.shared.actionHandler(doubleClickActionID, windowUUID: windowUUID, viewID: element.id, viewPartID: 0, context: index)
            }
        }
        #else
        return SwiftUI.EmptyView()
        #endif
    }
    
    static var initialValue: (ViewModel) -> Any? = { model in
        if let initialValue = model.value as? [String] {
            return initialValue
        }
        return [] as [String]
    }
}

File: ActionUIWeb/src/Views/Table.js
// Table.js — Table element.
// Web analog of ActionUI/Views/Table.swift (and ActionUIAndroid Views/Table.kt).
//
// A multi-column data table. Apple's Table is macOS-only (SwiftUI.Table lives on
// AppKit; every other Apple platform returns EmptyView), and Android — the "every
// other platform" case — renders nothing. The web is a desktop target like macOS,
// so it renders a real <table>, following the macOS reference rather than the
// Android no-op.
//
// Table is purely data-driven: it has no static (children) mode. Rows come from
// states["content"] ([[String]]) via the rows API on Window
// (set/append/clearElementRows, getElementColumnCount) — landed alongside this
// element, since Table is the first consumer that needs it. The Table binds its
// "content" state so a set/append/clear re-renders the body in place.
//
// Selection (value bridge). Like List, the selected row is carried as a String:
// its columns tab-joined ("" = nothing selected), read/written with
// get/setString(tableID) — Android's string transport for the Apple [String]
// value. A row can also be selected programmatically (without replacing the rows,
// firing no actionID) via the model selection API — selectElementRow(index) /
// selectElementRowWithContent(text, column) / clearElementSelection.
// The table-level actionID fires on selection change (no context; the host
// reads the value, matching Apple). doubleClickActionID fires on a double-click of
// the selected row with the row index as context. A Button cell fires its own
// columnTypes[c].actionID and consumes the click, so per-cell actions and row
// selection coexist.
//
// Properties (mirroring Table.swift):
//   columns          Required [String] of header titles.
//   columnHeadersVisibility  "automatic"|"hidden"|"visible".
//   columnTypes      Per-column { viewType, actionContext, actionID,
//                    dataInterpretation }; padded to columns with Text.
//   widths / minWidths  [Int] ideal / minimum column widths; the widest-ideal
//                    column auto-sizes to fill the frame (macOS), the rest take
//                    their ideal. Every column is user-resizable by dragging its
//                    header border (down to the resolved minimum); the table is
//                    wrapped in a scroller, so resizing past the frame (or adding
//                    rows) scrolls rather than clips. Resizing needs visible headers.
//   actionID / doubleClickActionID  selection / double-click actions.

import { register } from "../Common/ActionUIRegistry.js";
import { markHandlesAction } from "../Common/ModifierResolver.js";
import { buildDataImageCell } from "../Helpers/DataImageCell.js";
import { commonRowPrefix } from "../Helpers/RowDiff.js";

const CELL_VIEW_TYPES = ["Text", "Button", "Image", "AsyncImage"];
const DATA_INTERPRETATIONS = ["path", "systemName", "assetName", "resourceName", "mixed"];
const CELL_ACTION_CONTEXTS = ["title", "rowIndex", "columnIndex", "rowColumnIndex"];
const HEADER_VISIBILITIES = ["visible", "hidden", "automatic"];

register("Table", {
    // The selected row, columns tab-joined, as a String ("" = none). See header.
    valueType: "string",

    // Mirrors Table.swift validateProperties (warning text included verbatim).
    validateProperties: (properties, logger) => {
        const validated = { ...properties };

        // columns — an array of strings, else []. (Apple writes back [] when nil
        // or wrong-typed; only the wrong-typed case warns.)
        if (validated.columns === undefined) {
            validated.columns = [];
        } else if (!isStringArray(validated.columns)) {
            logger.log("Table columns must be an array of strings; defaulting to []", "warning");
            validated.columns = [];
        }

        // columnTypes — pad to columns.length with Text, then validate each entry.
        const columnTypes = Array.isArray(properties.columnTypes)
            ? properties.columnTypes.map((entry) => ({ ...entry })) : [];
        while (columnTypes.length < validated.columns.length) {
            columnTypes.push({ viewType: "Text" });
        }
        columnTypes.forEach((ct, i) => {
            const vt = typeof ct.viewType === "string" ? ct.viewType : "Text";
            if (!CELL_VIEW_TYPES.includes(vt)) {
                logger.log(`Table columnTypes[${i}].viewType must be 'Text', 'Button', 'Image', or 'AsyncImage'; defaulting to Text`, "warning");
                ct.viewType = "Text";
            }
            if (vt === "Image" && !DATA_INTERPRETATIONS.includes(ct.dataInterpretation)) {
                logger.log(`Table columnTypes[${i}].dataInterpretation must be 'path', 'systemName', 'assetName', 'resourceName', or 'mixed' for Image; defaulting to systemName`, "warning");
                ct.dataInterpretation = "systemName";
            }
            if (vt === "Button") {
                if (!CELL_ACTION_CONTEXTS.includes(ct.actionContext)) {
                    logger.log(`Table columnTypes[${i}].actionContext must be 'title', 'rowIndex', 'columnIndex', or 'rowColumnIndex' for Button; defaulting to title`, "warning");
                    ct.actionContext = "title";
                }
                if (ct.dataInterpretation !== undefined && !DATA_INTERPRETATIONS.includes(ct.dataInterpretation)) {
                    logger.log(`Table columnTypes[${i}].dataInterpretation must be 'path', 'systemName', 'assetName', 'resourceName', or 'mixed' for Button; ignoring`, "warning");
                    delete ct.dataInterpretation;
                }
            }
        });
        validated.columnTypes = columnTypes;

        // widths / minWidths — arrays of integers, else dropped.
        if (validated.widths !== undefined && !isIntArray(validated.widths)) {
            logger.log("Table widths must be an array of integers; ignoring", "warning");
            delete validated.widths;
        }
        if (validated.minWidths !== undefined && !isIntArray(validated.minWidths)) {
            logger.log("Table minWidths must be an array of integers; ignoring", "warning");
            delete validated.minWidths;
        }

        // columnHeadersVisibility — a String in the valid set, else dropped.
        if (validated.columnHeadersVisibility !== undefined) {
            const chv = validated.columnHeadersVisibility;
            if (typeof chv === "string") {
                if (!HEADER_VISIBILITIES.includes(chv)) {
                    logger.log(`Invalid columnHeadersVisibility '${chv}'; expected one of ["visible", "hidden", "automatic"], ignoring`, "warning");
                    delete validated.columnHeadersVisibility;
                }
            } else {
                logger.log(`Invalid type for columnHeadersVisibility: expected String, got ${typeof chv}, ignoring`, "warning");
                delete validated.columnHeadersVisibility;
            }
        }

        // doubleClickActionID — must be a string, else dropped.
        if (validated.doubleClickActionID !== undefined && typeof validated.doubleClickActionID !== "string") {
            logger.log("Table doubleClickActionID must be a string; ignoring", "warning");
            delete validated.doubleClickActionID;
        }

        return validated;
    },

    initialValue: () => "",
    initialStates: () => ({ content: [] }),

    buildView: (element, properties, ctx) => {
        const columns = Array.isArray(properties.columns) ? properties.columns : [];
        const columnTypes = Array.isArray(properties.columnTypes) ? properties.columnTypes : [];
        const widths = Array.isArray(properties.widths) ? properties.widths : null;
        const minWidths = Array.isArray(properties.minWidths) ? properties.minWidths : null;
        // columnHeadersVisibility: "hidden" suppresses the header row entirely;
        // "visible" / "automatic" show it. A static (build-time) choice, matching
        // Apple/Android - there is no runtime toggle.
        const headersHidden = properties.columnHeadersVisibility === "hidden";

        // A scroll container wraps the table so it can grow past its frame: when the
        // columns are wider, or the rows taller, than the bounded frame, the wrapper
        // scrolls (horizontally / vertically) instead of clipping. The frame and the
        // data-aui-id land on this wrapper (it is the returned node); the header is
        // sticky so it stays visible while the rows scroll. (macOS SwiftUI.Table is
        // itself scroll-backed; this is the web equivalent.)
        const scroll = document.createElement("div");
        scroll.className = "aui-table-scroll";
        markHandlesAction(scroll); // selection owns the table's actionID (opt out of the generic click handler)

        const table = document.createElement("table");
        table.className = "aui-table";
        scroll.appendChild(table);

        // Column sizing. The widest-ideal column auto-sizes to fill the frame width
        // (matching macOS SwiftUI.Table); the others take their ideal width. Every
        // column is user-resizable by dragging its right-edge header handle.
        // `.aui-table` is `table-layout: fixed`, so the <col> widths are authoritative
        // and a drag is honored exactly.
        //
        // Widths are JS-managed (colWidths) rather than left to CSS so we can have BOTH
        // an auto-filling widest column AND a resize that pushes the table past the
        // frame (CSS alone gives one or the other: a width:100% table absorbs into its
        // auto column and never overflows). A ResizeObserver re-fills the widest column
        // when the frame changes; a drag sets the dragged column's width and the
        // wrapper scrolls once the total exceeds the frame. Dragging during a frame
        // that has room does not shrink other columns (the divider tracks the cursor) -
        // a deliberate, more predictable divergence from macOS' live column absorb.
        const idealWidths = columns.map((_, i) => (widths && Number.isInteger(widths[i])) ? widths[i] : 100);
        // Resolved minimum per column: an explicit minWidths entry, else min(ideal, 10)
        // (the Table.swift default). The floor a drag cannot cross.
        const resolvedMins = columns.map((_, i) =>
            (minWidths && Number.isInteger(minWidths[i])) ? minWidths[i] : Math.min(idealWidths[i], 10));
        // The widest-ideal column flexes (auto-fills the frame); first wins a tie.
        const maxIdeal = idealWidths.length ? Math.max(...idealWidths) : 0;
        const flexIndex = idealWidths.indexOf(maxIdeal);
        // Live width per column. Starts at the ideal (floored at the resolved minimum);
        // relayout() widens the flex column to fill once the frame is measured.
        const colWidths = idealWidths.map((w, i) => Math.max(w, resolvedMins[i]));
        // Set true when the user drags the flex column itself: it then keeps that manual
        // width and stops auto-filling (the host took control of the widest column).
        let flexPinned = false;

        const colEls = [];
        const colgroup = document.createElement("colgroup");
        columns.forEach((_, i) => {
            const col = document.createElement("col");
            if (minWidths && Number.isInteger(minWidths[i])) col.style.minWidth = `${minWidths[i]}px`;
            colgroup.appendChild(col);
            colEls.push(col);
        });
        table.appendChild(colgroup);

        // Push the live widths to the <col>s and size the table to their sum, so the
        // table is exactly as wide as its columns - the wrapper scrolls when that sum
        // exceeds the frame.
        const applyWidths = () => {
            let sum = 0;
            colWidths.forEach((w, i) => { colEls[i].style.width = `${w}px`; sum += w; });
            table.style.width = `${sum}px`;
        };

        // Re-fill the flex (widest) column so the table fills the frame width. `shrink`
        // lets it shrink back toward its ideal (used on a frame resize); without it the
        // flex column only grows to fill freed space (used on drag release), so
        // widening another column scrolls rather than silently shrinking the widest.
        const relayout = ({ shrink } = {}) => {
            const avail = scroll.clientWidth;
            if (flexIndex >= 0 && !flexPinned && avail > 0) {
                const othersSum = colWidths.reduce((s, w, i) => (i === flexIndex ? s : s + w), 0);
                const fill = Math.max(idealWidths[flexIndex], resolvedMins[flexIndex], avail - othersSum);
                colWidths[flexIndex] = shrink ? fill : Math.max(colWidths[flexIndex], fill);
            }
            applyWidths();
        };

        // Drag a column's right-edge handle to resize it. Pointer capture keeps the
        // drag tracking even when the pointer leaves the thin handle. The dragged
        // column tracks the cursor and the wrapper scrolls when the table outgrows the
        // frame; dragging the flex column pins it, and on release the flex column grows
        // to fill any freed space. (setPointerCapture / releasePointerCapture are
        // guarded so the headless test DOM still drives the move/up path.)
        const startResize = (colIndex, handle, event) => {
            event.preventDefault?.();
            event.stopPropagation?.(); // a header drag, never a row selection
            const startX = event.clientX;
            const startWidth = colWidths[colIndex];
            if (colIndex === flexIndex) flexPinned = true;
            handle.setPointerCapture?.(event.pointerId);
            scroll.classList.add("aui-table-resizing");
            const onMove = (moveEvent) => {
                colWidths[colIndex] = Math.max(resolvedMins[colIndex], startWidth + (moveEvent.clientX - startX));
                applyWidths();
            };
            const onUp = (upEvent) => {
                handle.releasePointerCapture?.(upEvent.pointerId);
                scroll.classList.remove("aui-table-resizing");
                handle.removeEventListener("pointermove", onMove);
                handle.removeEventListener("pointerup", onUp);
                relayout(); // grow the widest column to fill any freed space
            };
            handle.addEventListener("pointermove", onMove);
            handle.addEventListener("pointerup", onUp);
        };

        // The header row (suppressed entirely when columnHeadersVisibility is
        // "hidden"). Every column gets a drag handle on its right edge (the widest
        // column too - dragging it pins its width). Resizing needs a visible header to
        // grab, so a hidden-header table is not resizable.
        if (!headersHidden) {
            const thead = document.createElement("thead");
            const headRow = document.createElement("tr");
            columns.forEach((name, i) => {
                const th = document.createElement("th");
                th.textContent = String(name);
                const handle = document.createElement("div");
                handle.className = "aui-table-resizer";
                handle.setAttribute("aria-hidden", "true");
                handle.addEventListener("pointerdown", (event) => startResize(i, handle, event));
                th.appendChild(handle);
                headRow.appendChild(th);
            });
            thead.appendChild(headRow);
            table.appendChild(thead);
        }

        const tbody = document.createElement("tbody");
        table.appendChild(tbody);

        let rows = [];
        let trNodes = [];
        let selectedIndex = -1;

        const rowValue = () =>
            (selectedIndex >= 0 && selectedIndex < rows.length) ? rows[selectedIndex].join("\t") : "";

        const applySelectionStyles = () => {
            trNodes.forEach((tr, i) => {
                const selected = i === selectedIndex;
                tr.classList.toggle("aui-table-row-selected", selected);
                tr.setAttribute("aria-selected", selected ? "true" : "false");
            });
        };

        const selectIndex = (index, fromUser) => {
            if (index === selectedIndex) return;
            selectedIndex = index;
            applySelectionStyles();
            // Apple fires the selection action with no context and reads
            // model.value; the web matches — the selection is the element value.
            if (fromUser && typeof properties.actionID === "string") {
                ctx.model.dispatchAction(properties.actionID, element.id, 0, null);
            }
        };

        const buildCell = (value, colIndex, rowIndexRef) => {
            const td = document.createElement("td");
            const colType = columnTypes[colIndex] ?? { viewType: "Text" };
            const viewType = colType.viewType ?? "Text";
            if (viewType === "Button") {
                const button = document.createElement("button");
                button.className = "aui-table-button";
                if (colType.dataInterpretation) {
                    button.appendChild(buildDataImageCell(value, colType.dataInterpretation, ctx.logger, "Table"));
                } else {
                    button.textContent = value;
                }
                button.addEventListener("click", (event) => {
                    event.stopPropagation(); // the cell's own tap, not row selection
                    if (typeof colType.actionID === "string") {
                        ctx.model.dispatchAction(colType.actionID, element.id, colIndex, cellContext(colType.actionContext, value, rowIndexRef(), colIndex));
                    }
                });
                td.appendChild(button);
            } else if (viewType === "Image") {
                td.appendChild(buildDataImageCell(value, colType.dataInterpretation ?? "mixed", ctx.logger, "Table"));
            } else if (viewType === "AsyncImage") {
                const img = document.createElement("img");
                img.className = "aui-data-image";
                img.loading = "lazy";
                img.src = value;
                img.alt = "";
                td.appendChild(img);
            } else { // Text and fallback
                td.textContent = value;
            }
            return td;
        };

        const buildRow = (row, rowIndex) => {
            const tr = document.createElement("tr");
            tr.setAttribute("role", "row");
            const rowIndexRef = () => rowIndex;
            columns.forEach((_, colIndex) => {
                tr.appendChild(buildCell(row[colIndex] ?? "", colIndex, rowIndexRef));
            });
            tr.addEventListener("click", () => selectIndex(rowIndex, true));
            tr.addEventListener("dblclick", () => {
                if (typeof properties.doubleClickActionID === "string" && rowIndex === selectedIndex) {
                    ctx.model.dispatchAction(properties.doubleClickActionID, element.id, 0, rowIndex);
                }
            });
            return tr;
        };

        // Apply a new rows array with a common-prefix diff (Helpers/RowDiff.js)
        // instead of rebuilding the whole <tbody>: an append (the rows API re-sends
        // the whole array) keeps every unchanged prefix <tr> - its cells, selection
        // styling and baked row index - in place and builds only the new tail; a
        // prepend / mid-list edit rebuilds from the first change. Column auto-fill is
        // width-based (row-count independent), so an append needs no relayout.
        const applyRows = (next) => {
            next = Array.isArray(next) ? next : [];
            const keep = commonRowPrefix(rows, next);
            while (trNodes.length > keep) { trNodes[trNodes.length - 1].remove(); trNodes.pop(); }
            for (let rowIndex = keep; rowIndex < next.length; rowIndex++) {
                const tr = buildRow(next[rowIndex], rowIndex);
                trNodes.push(tr);
                tbody.appendChild(tr);
            }
            rows = next;
            // A re-render drops the prior selection if its row no longer exists.
            if (selectedIndex >= rows.length) selectedIndex = -1;
            applySelectionStyles();
        };

        if (element.id > 0) {
            // The rows store: set/append/clearElementRows route through here and
            // re-render the body in place (incrementally - see applyRows).
            ctx.model.bindState(element.id, {
                getState: (key) => (key === "content" ? rows : undefined),
                setState: (key, value) => { if (key === "content") applyRows(value); },
            });
            // The selection value: tab-joined selected row; set by value match.
            ctx.model.bind(element.id, {
                getValue: () => rowValue(),
                setValue: (value) => { // programmatic: silent
                    const target = String(value ?? "");
                    selectedIndex = target === "" ? -1 : rows.findIndex((row) => row.join("\t") === target);
                    applySelectionStyles();
                },
            });
        }

        applyWidths(); // seed the <col> widths before the frame is measured
        applyRows([]);
        // Keep the widest column filling the frame as the frame / window resizes
        // (also performs the first fill once the wrapper has a measured width).
        if (typeof ResizeObserver !== "undefined") {
            const observer = new ResizeObserver(() => relayout({ shrink: true }));
            observer.observe(scroll);
        }
        return scroll;
    },
});

// The Button cell's action context, per columnTypes[c].actionContext.
function cellContext(actionContext, value, rowIndex, colIndex) {
    switch (actionContext) {
        case "rowIndex":       return rowIndex;
        case "columnIndex":    return colIndex;
        case "rowColumnIndex": return { row: rowIndex, column: colIndex };
        default:               return value; // "title"
    }
}

function isStringArray(value) {
    return Array.isArray(value) && value.every((entry) => typeof entry === "string");
}

function isIntArray(value) {
    return Array.isArray(value) && value.every((entry) => Number.isInteger(entry));
}
